src/mlpack/methods/ann/image_functions/bilinear_function.hpp

```diff
@@ -0,0 +1,187 @@
+/**
+ * @file bilinear_function.hpp
+ * @author Kris Singh
+ *
+ * Definition and implementation of the bilinear interpolation function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_BILINEAR_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_BILINEAR_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <assert.h>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The bilinear interpolation function
+ *
+ */
+class BiLinearFunction
+{
+ /*
+  * The constructor for the bilinear interpolation.
+  * 
+  * @param inRowSize: Number of input rows
+  * @param inColSize: Number of input columns.
+  * @param outRowSize: Number of output rows.
+  * @param outColSie: Number of output columns.
+  */
+ public:
+  BiLinearFunction(
+      const size_t inRowSize,
+      const size_t inColSize,
+      const size_t outRowSize,
+      const size_t outColSize):
+      inRowSize(inRowSize),
+      inColSize(inColSize),
+      outRowSize(outRowSize),
+      outColSize(outColSize)
+  {};
+  /**
+   * UpSample the given input.
+   *
+   * @param input The input matrix
+   * @param output The resulting interpolated output image.
+   */
+  template<typename eT>
+  void UpSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    if (output.is_empty())
+      output.set_size(outRowSize * outColSize, 1);
+
+    assert(output.n_rows == outRowSize * outColSize);
+    assert(output.n_cols == 1);
+
+    scaleRow = (double)inRowSize / (double) outRowSize;
+    scaleCol = (double)inColSize / (double) outColSize;
+
+    for (size_t i = 0; i < outRowSize; i++)
+    {
+      for (size_t j = 0; j < outColSize; j++)
+      {
+        rOrigin = std::floor(i * scaleRow);
+        cOrigin = std::floor(j * scaleCol);
+
+        if (rOrigin > inRowSize - 2)
+          rOrigin = inRowSize - 2;
+        if (cOrigin > inColSize - 2)
+          cOrigin = inColSize - 2;
+        
+
+        double deltaR = i * scaleRow - rOrigin;
+        double deltaC = j * scaleCol - cOrigin;
+        coeff1 = (1 - deltaR) * (1 - deltaC);
+        coeff2 = deltaR * (1 - deltaC);
+        coeff3 = (1 - deltaR) * deltaC;
+        coeff4 = deltaR * deltaC;
+
+        output(i * outRowSize + j) =  
```

Comment:
Probably, it is better to store matrices in column-major order
since Armadillo does the same i.e.
`output(i, j) = output(j * outColSize + i)`
and
`input(a, b) = input(b * inColSize + a)`